### ISSUE TITLE
[Bug] Original and rendered image pixels don't match although canvas size matches original image size

### ISSUE THREAD

**Issue by canon-cmre-ben-layet** [cornerstonejs/cornerstone3D#1480]
### Describe the Bug

I'm doing a simple 2D render with cornerstone and using setDisplayArea to set an imageArea of [1,1].
If the canvas created by cornerstone has the same size as original data then I expect the pixels to match 1:1 between original and rendered images. Currently there is a 1 pixel x and y offset at top left of rendered image.

### Steps to Reproduce

Please see the jsfiddle I made:  https://jsfiddle.net/t839Lgzr/33/
You can right click on rendered image and "copy image" and save it and compare with the original "grid.png" I attached here. I also attached an image I obtained by doing the copy image: "jsfiddle-canvas-256-dpr-2.png"
Also you can compare with the DICOM source file used in the example:
https://canon-cmre-ben-layet.github.io/shared-files/dicom-grid.dcm

**NOTE**: this example relies on the image size (512x512) dividing nicely by browser device pixel ratio - to give a whole number. Only in this case will the cornerstone created canvas be exactly 512x512, and we can have simple expectations about what the rendered images should look like. Using a display scaling of 100% or 200% shows the issue.

"jsfiddle-canvas-256-dpr-2.png"
![jsfiddle-canvas-256-dpr-2](https://github.com/user-attachments/assets/02854508-4f1b-486a-a3f0-96e4f4d02e7b)
original "grid.png"
![grid](https://github.com/user-attachments/assets/53668b89-a6a4-494e-b417-82cc01af920a)

### The current behavior

Currently the rendered image has a 1 pixel offset from what is expected at the top left.
(You can see the x=0 and y=0 black lines present in original are missing.)
At the bottom right the original and rendered images align.
In the middle the rendered image is fuzzy as the pixel alignment is off by half a pixel.
**NOTE**: As stated above, this description is only accurate for some device pixel ratios, such as of 1 or 2. But I assume that the issue is present in other cases - just harder to observe.

### The expected behavior

A 1:1 rendering of original image to a canvas of the same size should be seen with the conditions described.

**NOTE**: For some background: I would really like to be able to use cornerstone to do 1:1 renders of 2D data. So render 512x512 DICOM to a 512x512 canvas. I am not sure what the best approach is for this. I'll ask a question separately about this

### OS

Windows 10

### Node version

n/a

### Browser

Chrome

**Comment by canon-cmre-ben-layet** [comment]
Also note that add grid.dcm to an orthanc server and send this request to the DICOMweb /rendered API:
```
http://{{server}}/dicom-web/studies/1.2.276.0.7230010.3.1.4.3011000561.4200.1727173121.951/series/1.2.276.0.7230010.3.1.4.3011000561.4200.1727173121.952/instances/1.2.276.0.7230010.3.1.4.3011000561.4200.1727173121.950/rendered?viewport=512,512
```
Then the image you get back exactly matches the original "grid.png" I uploaded.

**Comment by sedghi** [comment]
This renders correctly in our local drag and drop 
![CleanShot 2024-11-14 at 16 23 47@2x](https://github.com/user-attachments/assets/fbb83ac7-54c1-4d36-9800-73472e3f9710)

https://www.cornerstonejs.org/live-examples/local.html

Can i close this ?

**Comment by canon-cmre-ben-layet** [comment]
Hi thanks for taking a look.
I would like you to look at the example that I prepared.
The rendering parameters are different from those used in the live-examples page.
The jsfiddle was broken due to some API changes. I updated it: https://jsfiddle.net/h37tjcgx/
It shows the same behaviour as before.
Please note that setting display scaling to 100%, 200%, (i.e. device pixel ratio to 1, 2,) is I think important in order to have a simple expectation about what is rendered. I find it easier to think about anyway.
Thanks!

**Comment by sedghi** [comment]
I think the issue is the `setDisplayArea` , if i uncomment it it looks correct to my eyes

<img src="https://uploads.linear.app/e0b640a9-3fd8-4513-bc84-af25921e5dc8/a6dd8974-96f7-40f4-953c-581ca4c39957/8430722b-7c80-47b5-9a3d-a70331d822d2?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiL2UwYjY0MGE5LTNmZDgtNDUxMy1iYzg0LWFmMjU5MjFlNWRjOC9hNmRkODk3NC05NmY3LTQwZjQtOTUzYy01ODFjYTRjMzk5NTcvODQzMDcyMmItN2M4MC00N2I1LTlhM2QtYTcwMzMxZDgyMmQyIiwiaWF0IjoxNzMxOTQyNTcyLCJleHAiOjMzMzAyNTAyNTcyfQ.30HfxPFq6alQ4nWjIEsh9M7WawWlYSJZtiOPghPewbc " alt="CleanShot 2024-11-18 at 10.08.47@2x.png" width="900" height="708" />

**Comment by canon-cmre-ben-layet** [comment]
I assume you mean if you comment out the setDisplayArea.
Sure it looks different then. Whether it's correct I don't know.
But anyway I want to call setDisplayArea. I don't want the border that you get without calling it.
I want to render the data to a 512x512 image.
In the configuration I want it looks like the rendered image isn't correct.

**Comment by sedghi** [comment]
Yeah i understand your comment, I will look at the setDisplayArea implementation

**Comment by sedghi** [comment]
can you try this branch https://github.com/cornerstonejs/cornerstone3D/pull/1701

**Comment by canon-cmre-ben-layet** [comment]
@sedghi how can i try the branch? is a built version available or should I clone and build the repo at this revision? please give details

**Comment by sedghi** [comment]
i was thinking just check out and run one of the examples with your data

**Comment by canon-cmre-ben-layet** [comment]
unless the example matches my usage that won't really prove anything. maybe I could modify an example. not sure.
are there instructions for building and running them?

**Comment by canon-cmre-ben-layet** [comment]
Alternatively the data I used is available here: https://canon-cmre-ben-layet.github.io/shared-files/dicom-grid.dcm
And this shows my usage: : https://jsfiddle.net/h37tjcgx/
If you wanted to do a quick test.

**Comment by sedghi** [comment]
you can try latest and see if it resolved

**Comment by canon-cmre-ben-layet** [comment]
Seems same behaviour to me. Rendered image blurring in middle, crisp at edges.
You can also test by opening attached file in browser (after renaming from .txt to .html).
Copy out image from page using right click menu. Paste in Paint or whatever to inspect.
[index-setdisplayarea.txt](https://github.com/user-attachments/files/18347110/index-setdisplayarea.txt)

**Comment by sedghi** [comment]
I didn't know the issue you are describing is for bluring in the middle and crisp on the edge, hmm, that seems to be an interpolation issue. How about you setProperties({ interpolationType } ) i guess the enum for it is 0 or might be 1

**Comment by canon-cmre-ben-layet** [comment]
I tried following in the index-setdisplayarea.html. Didn't see any behaviour changes for <type> of 0 or 1.
    viewport.setStack([gridImage]);
    viewport.setProperties({ interpolationType: <type> });
    viewport.setDisplayArea({ imageArea: [1, 1], storeAsInitialCamera: true });

**Comment by wayfarer3130** [comment]
You should set the display area to scale 1 the same as the example https://www.cornerstonejs.org/live-examples/resize
when you choose "scale 1"
The issue is that you need to get the exact size and position correctly adjusted so that you don't have fractional pixels, deal with screen scaling and a few other issues.
